pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -818,6 +821,34 @@ func (c *Controller) WorkloadInstanceHandler(si *model.WorkloadInstance, event m
 	default: // add or update
 		c.workloadInstancesByIP[si.Endpoint.Address] = si
 	}
+
+	// Update the config map for workload's service discovery as well.
+	// External workloads have no endpoint registered in k8s, so the traffic sent to these non-k8s workloads
+	// won't be picked up by Prometheus. We update the workload entries' endpoints to a config map so that
+	// Prometheus could mount the map and read the endpoints.
+	wleName := createNameFromAddr(si.Endpoint.Address)
+	fileSDConfig := c.GetOrCreatePromSDConfigMap()
+	switch event {
+	case model.EventDelete:
+		if fileSDConfig.Data != nil {
+			delete(fileSDConfig.Data, fmt.Sprintf("%s.yaml", wleName))
+		}
+	default: // add or update
+		if fileSDConfig.Data == nil {
+			fileSDConfig.Data = make(map[string]string)
+		}
+		staticConfig := `
+- targets:
+  - %s
+`
+		fileSDConfig.Data[fmt.Sprintf("%s.yaml", wleName)] = fmt.Sprintf(staticConfig, si.Endpoint.Address)
+	}
+	// Write the update config map back to cluster
+	if _, err := c.client.CoreV1().ConfigMaps("istio-system").Update(context.TODO(), fileSDConfig,
```

Comment:
There is a great possibility to race when there are multi-istiod

Reply:
Thanks for pointing this out! If we have multi-istiod, normally that's the case when we have multiple clusters? Then won't this line write to their own cluster? Or am I understanding it wrong? Thanks!

Reply:
Multi instances can be deployed in one cluster, for HA

Reply:
Understood! This is helpful for me. And this is just a proof of concept for now and we might go with different approaches here. Thanks for reviewing!